File: backend/app/schemas/analytics_dashboard.py
from pydantic import BaseModel


class ZoneTransition(BaseModel):
    from_zone_id: int
    from_zone_name: str
    to_zone_id: int
    to_zone_name: str
    count: int


class JourneyFlowResponse(BaseModel):
    store_id: int | None
    transitions: list[ZoneTransition]


class SegmentItem(BaseModel):
    segment: str
    count: int
    avg_dwell_seconds: float


class SegmentationResponse(BaseModel):
    store_id: int | None
    total_customers: int
    segments: list[SegmentItem]
    multi_zone_visitor_pct: float


class StoreComparisonItem(BaseModel):
    store_id: int
    store_name: str
    visitors: int
    avg_dwell_seconds: float
    peak_hour: int | None


class StoreComparisonResponse(BaseModel):
    stores: list[StoreComparisonItem]


class InsightItem(BaseModel):
    severity: str
    message: str


class InsightsResponse(BaseModel):
    store_id: int | None
    insights: list[InsightItem]


class ProductVisibilityItem(BaseModel):
    shelf_id: int
    shelf_name: str
    store_name: str
    zone: str
    visibility_score: int


class ProductVisibilityResponse(BaseModel):
    shelves: list[ProductVisibilityItem]


class FunnelStage(BaseModel):
    stage: str
    count: int


class ConversionFunnelResponse(BaseModel):
    store_id: int | None
    stages: list[FunnelStage]
    conversion_rate: float | None


class CategorySummaryItem(BaseModel):
    category: str
    product_count: int
    total_stock: int
    avg_price: float
    inventory_value: float


class InventorySummaryResponse(BaseModel):
    store_id: int | None
    total_products: int
    total_inventory_items: int
    inventory_value: float
    low_stock_products: int
    categories: list[CategorySummaryItem]


class ProductStockItem(BaseModel):
    product_id: int
    product_name: str
    sku: str
    category: str
    stock_quantity: int
    price: float
    shelf_name: str


class ProductAnalysisResponse(BaseModel):
    store_id: int | None
    highest_stock: list[ProductStockItem]
    lowest_stock: list[ProductStockItem]
    to_restock: list[ProductStockItem]
    categories: list[CategorySummaryItem]


class ShelfAnalysisItem(BaseModel):
    shelf_id: int
    shelf_name: str
    store_name: str
    zone: str
    product_count: int
    visit_count: int
    status: str
    # Live re-detection on the shelf camera's latest processed snapshot,
    # cross-checked against product_count (the DB's expected stock record
    # count). None if this camera has never been processed - nothing to
    # cross-check against yet, not "detected zero".
    detected_product_count: int | None = None
    restocking_needed: bool = False


class ShelfAnalysisResponse(BaseModel):
    store_id: int | None
    shelves: list[ShelfAnalysisItem]
    most_visited: ShelfAnalysisItem | None
    least_visited: ShelfAnalysisItem | None
    empty_count: int
    full_count: int


class AttractivenessItem(BaseModel):
    shelf_id: int
    shelf_name: str
    store_name: str
    zone: str
    score: float
    traffic_score: float
    dwell_score: float
    interaction_score: float
    stockout_penalty: float
    rank: int
    has_behavior_data: bool


class AttractivenessResponse(BaseModel):
    store_id: int | None
    shelves: list[AttractivenessItem]


class RecommendationItem(BaseModel):
    severity: str
    shelf_id: int
    shelf_name: str
    zone: str
    issue: str
    action: str


class RecommendationsResponse(BaseModel):
    store_id: int | None
    recommendations: list[RecommendationItem]
    # Shelves that actually had tracking data to rank against - lets the
    # frontend show an honest "not enough data yet" state instead of an
    # empty list that looks like "everything's fine".
    shelves_considered: int

File: frontend/src/api/analyticsDashboard.ts
import api from "./client";

export interface ZoneTransition {
  from_zone_id: number;
  from_zone_name: string;
  to_zone_id: number;
  to_zone_name: string;
  count: number;
}
export interface JourneyFlowResponse {
  store_id: number | null;
  transitions: ZoneTransition[];
}

export interface SegmentItem {
  segment: string;
  count: number;
  avg_dwell_seconds: number;
}
export interface SegmentationResponse {
  store_id: number | null;
  total_customers: number;
  segments: SegmentItem[];
  multi_zone_visitor_pct: number;
}

export interface StoreComparisonItem {
  store_id: number;
  store_name: string;
  visitors: number;
  avg_dwell_seconds: number;
  peak_hour: number | null;
}
export interface StoreComparisonResponse {
  stores: StoreComparisonItem[];
}

export interface InsightItem {
  severity: string;
  message: string;
}
export interface InsightsResponse {
  store_id: number | null;
  insights: InsightItem[];
}

export interface ProductVisibilityItem {
  shelf_id: number;
  shelf_name: string;
  store_name: string;
  zone: string;
  visibility_score: number;
}
export interface ProductVisibilityResponse {
  shelves: ProductVisibilityItem[];
}

export interface FunnelStage {
  stage: string;
  count: number;
}
export interface ConversionFunnelResponse {
  store_id: number | null;
  stages: FunnelStage[];
  conversion_rate: number | null;
}

export interface CategorySummaryItem {
  category: string;
  product_count: number;
  total_stock: number;
  avg_price: number;
  inventory_value: number;
}

export interface InventorySummaryResponse {
  store_id: number | null;
  total_products: number;
  total_inventory_items: number;
  inventory_value: number;
  low_stock_products: number;
  categories: CategorySummaryItem[];
}

export interface ProductStockItem {
  product_id: number;
  product_name: string;
  sku: string;
  category: string;
  stock_quantity: number;
  price: number;
  shelf_name: string;
}

export interface ProductAnalysisResponse {
  store_id: number | null;
  highest_stock: ProductStockItem[];
  lowest_stock: ProductStockItem[];
  to_restock: ProductStockItem[];
  categories: CategorySummaryItem[];
}

export interface ShelfAnalysisItem {
  shelf_id: number;
  shelf_name: string;
  store_name: string;
  zone: string;
  product_count: number;
  visit_count: number;
  status: string;
  /** Live re-detection on this shelf camera's latest processed frame.
   *  null when the camera has never been processed - which is different from
   *  "detected zero products". */
  detected_product_count: number | null;
  restocking_needed: boolean;
}

export interface ShelfAnalysisResponse {
  store_id: number | null;
  shelves: ShelfAnalysisItem[];
  most_visited: ShelfAnalysisItem | null;
  least_visited: ShelfAnalysisItem | null;
  empty_count: number;
  full_count: number;
}

export interface AttractivenessItem {
  shelf_id: number;
  shelf_name: string;
  store_name: string;
  zone: string;
  score: number;
  traffic_score: number;
  dwell_score: number;
  interaction_score: number;
  stockout_penalty: number;
  rank: number;
  has_behavior_data: boolean;
}

export interface AttractivenessResponse {
  store_id: number | null;
  shelves: AttractivenessItem[];
}

export interface RecommendationItem {
  severity: string;
  shelf_id: number;
  shelf_name: string;
  zone: string;
  issue: string;
  action: string;
}

export interface RecommendationsResponse {
  store_id: number | null;
  recommendations: RecommendationItem[];
  shelves_considered: number;
}

function storeParams(storeId?: number) {
  return storeId ? { store_id: storeId } : {};
}

export const analyticsDashboardApi = {
  journeyFlow: (storeId?: number) =>
    api.get<JourneyFlowResponse>("/dashboard/analyst/journey-flow", { params: storeParams(storeId) }),
  segmentation: (storeId?: number) =>
    api.get<SegmentationResponse>("/dashboard/analyst/segmentation", { params: storeParams(storeId) }),
  storeComparison: () => api.get<StoreComparisonResponse>("/dashboard/analyst/store-comparison"),
  insights: (storeId?: number) =>
    api.get<InsightsResponse>("/dashboard/analyst/insights", { params: storeParams(storeId) }),
  productVisibility: (storeId?: number) =>
    api.get<ProductVisibilityResponse>("/dashboard/marketing/product-visibility", { params: storeParams(storeId) }),
  conversionFunnel: (storeId?: number) =>
    api.get<ConversionFunnelResponse>("/dashboard/marketing/conversion-funnel", { params: storeParams(storeId) }),
  inventorySummary: (storeId?: number) =>
    api.get<InventorySummaryResponse>("/dashboard/analyst/inventory-summary", { params: storeParams(storeId) }),
  productAnalysis: (storeId?: number) =>
    api.get<ProductAnalysisResponse>("/dashboard/analyst/product-analysis", { params: storeParams(storeId) }),
  shelfAnalysis: (storeId?: number) =>
    api.get<ShelfAnalysisResponse>("/dashboard/analyst/shelf-analysis", { params: storeParams(storeId) }),
  attractiveness: (storeId?: number) =>
    api.get<AttractivenessResponse>("/dashboard/attractiveness", { params: storeParams(storeId) }),
  recommendations: (storeId?: number) =>
    api.get<RecommendationsResponse>("/dashboard/recommendations", { params: storeParams(storeId) }),
};
